### PR TITLE
DevDocs: Generate an index of proptypes for client consumption

### PR DESCRIPTION
**Pull request by withinboredom**
One of the first steps in rendering the propTypes in the client is getting them to the client. This creates a new artifact during a build (194k -- 24k gzip'd) that parses **all** the `index.jsx` and `example.jsx` files. The reasoning behind this is two fold:

1. We don't really know at build time which components are in the devdocs. While we could discover them, it would add complexity.
2. We also want the `example.jsx` files for now and they will be removed in a future iteration. This is so that a user of this file can know what props are exposed through the example wrapper component, if any.

I believe this PR opens the door to some really awesome things. Such as:

1. Searching for a component by includePath
2. Adding a props reference to readme's as part of the build
3. Adding a props reference in devdocs

## The generated file

An array of parsed objects, which looks something like this for a small component:

``` json
{
	"description": "",
	"displayName": "BulkSelect",
	"methods": [
	  {
		"name": "getStateIcon",
		"docblock": null,
		"modifiers": [],
		"params": [],
		"returns": null
	  },
	  {
		"name": "hasAllElementsSelected",
		"docblock": null,
		"modifiers": [],
		"params": [],
		"returns": null
	  },
	  {
		"name": "hasSomeElementsSelected",
		"docblock": null,
		"modifiers": [],
		"params": [],
		"returns": null
	  },
	  {
		"name": "handleToggleAll",
		"docblock": null,
		"modifiers": [],
		"params": [],
		"returns": null
	  }
	],
	"props": {
	  "totalElements": {
		"type": {
		  "name": "number"
		},
		"required": true,
		"description": ""
	  },
	  "selectedElements": {
		"type": {
		  "name": "number"
		},
		"required": true,
		"description": ""
	  },
	  "onToggle": {
		"type": {
		  "name": "func"
		},
		"required": true,
		"description": ""
	  }
	},
	"includePath": "components/bulk-select",
	"slug": "bulk-select"
  },
```

### CODE REVIEW

**Comment by ehg** [comment]
Cool.

As discussed on Slack, I think we could get rid of the async functions, and go almost completely synchronous here.

**Comment by ehg** [comment]
> This has several sub-properties allowing a user to easily find components without having to filter the entire collection of data. Each key is the index and each value is an array of numbers that correspond to the index in the data array.

Are we pre-optimising a bit here? What are the advantages of putting these indices in the data structure, rather than having a bunch of filtering methods available?

File: npm-shrinkwrap.json [hunk 1]
```diff
@@ -4029,7 +4058,7 @@
       }
     },
     "webpack-core": {
-      "version": "0.6.8",
+      "version": "0.6.9",
```
Comment (withinboredom) [inline npm-shrinkwrap.json]: It's this guy: https://github.com/mzgoddard/hard-source-webpack-plugin/blob/master/package.json#L50

File: server/devdocs/bin/generate-proptypes-index.js [hunk 2]
```diff
@@ -0,0 +1,198 @@
+#!/usr/bin/env node
```
Comment (withinboredom) [inline server/devdocs/bin/generate-proptypes-index.js]: Fixed, weird that it didn't catch the exec bit when I originally committed it

**Comment by withinboredom** [comment]
I've made the processing of the files synchronous, the amount of time it takes is the same as asynchronous 😄

**Comment by withinboredom** [comment]
> Are we pre-optimising a bit here?

Just a little 😉 

> What are the advantages of putting these indices in the data structure, rather than having a bunch of filtering methods available?

I knew from #9289 that having those would make the client code a tad simpler (by removing any reducing), but removing them from this code makes this code much much simpler!

File: Makefile [hunk 3]
```diff
@@ -59,6 +60,9 @@ COMPONENTS_USAGE_STATS_FILES = $(shell \
 		-type f \
 		\( -name '*.js' -or -name '*.jsx' \) \
 )
+COMPONENTS_PROPTYPE_FILES = $(shell \
+	find client -name 'index.jsx' -or -name 'example.jsx' \
```
Comment (ehg) [inline Makefile]: We sometimes use the `.js` extension, even if the component contains JSX. e.g. `components/signup-site-title/index.js`, we should probably account for these, naming philosophy aside. We'll probably want to ignore `test/` files as well.

Also, what do you think about limiting the files we scan to `blocks` and `components`?

File: package.json [hunk 4]
```diff
@@ -98,6 +98,7 @@
     "react-addons-update": "15.4.0",
     "react-click-outside": "2.1.0",
     "react-day-picker": "2.4.1",
+    "react-docgen": "2.12.1",
```
Comment (ehg) [inline package.json]: Can we bump to the version without `nomnom` now? :)

File: server/devdocs/bin/generate-proptypes-index.js [hunk 5]
```diff
@@ -0,0 +1,124 @@
...
+
+/**
+ * This file generates an index of proptypes by component displayname, slug and folder name
+ */
+
+const fs = require( 'fs' ),
```
Comment (ehg) [inline server/devdocs/bin/generate-proptypes-index.js]: We prefer separate `consts` per line now, as per `imports`

Comment (ehg) [inline server/devdocs/bin/generate-proptypes-index.js]: Also, let's add an external dependencies docblock, and keep `root` and `pathSwap` separate.

File: server/devdocs/bin/generate-proptypes-index.js [hunk 6]
```diff
@@ -0,0 +1,124 @@
...
+	path = require( 'path' ),
+	async = require( 'async' ),
```
Comment (ehg) [inline server/devdocs/bin/generate-proptypes-index.js]: Do we need this module now?

File: Makefile (hunk 3 above)
Comment (withinboredom) [inline Makefile]: There's some components in the devdocs that are outside of the components and blocks folders:

- my-sites/post-selector/docs/example
- lib/sites-list/docs/example
- my-sites/plan-storage/docs/example

for example ...

File: package.json (hunk 4 above)
Comment (withinboredom) [inline package.json]: Yeah, just released last night.

File: server/devdocs/bin/generate-proptypes-index.js (hunk 6 above)
Comment (withinboredom) [inline server/devdocs/bin/generate-proptypes-index.js]: :facepalm:

File: Makefile (hunk 3 above)
Comment (ehg) [inline Makefile]: > There's some components in the devdocs that are outside of the components and blocks folders

Indeed. But those examples should probably be [blocks](https://github.com/Automattic/wp-calypso/tree/master/client/blocks#blocks) — apart from `sites-list`, which isn't a component at all, and should not exist :)

Comment (ehg) [inline Makefile]: Actually, I think having data on components outside `blocks` and `components` will be useful for future README generation. So ignore this :)

**Comment by ehg** [comment]
LGTM. Tested building the Docker image, works fine.

**Pull request merged**